tools/apprunner-common/src/test/java/org/projectnessie/quarkus/runner/TestProcessHandler.java

```diff
@@ -0,0 +1,315 @@
+/*
+ * Copyright (C) 2020 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.quarkus.runner;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.RepeatedTest;
+import org.junit.jupiter.api.Test;
+
+class TestProcessHandler {
+
+  private static ExecutorService executor;
+
+  @BeforeAll
+  static void createExecutor() {
+    executor = Executors.newCachedThreadPool();
+  }
+
+  @SuppressWarnings("ResultOfMethodCallIgnored")
+  @AfterAll
+  static void stopExecutor() throws Exception {
+    executor.shutdown();
+    executor.awaitTermination(10, TimeUnit.SECONDS);
+  }
+
+  @Test
+  void notStarted() {
+    ProcessHandlerMock phMock = new ProcessHandlerMock();
+
+    assertThatThrownBy(phMock.ph::stop)
+        .isInstanceOf(IllegalStateException.class)
+        .hasMessage("No process started");
+  }
+
+  @Test
+  void doubleStart() {
+    ProcessHandlerMock phMock = new ProcessHandlerMock();
+
+    phMock.ph.started(phMock.proc);
+
+    assertThatThrownBy(() -> phMock.ph.started(phMock.proc))
+        .isInstanceOf(IllegalStateException.class)
+        .hasMessage("Process already started");
+  }
+
+  @RepeatedTest(20) // repeat, risk of flakiness
```

Comment:
why are they flaky?

Reply:
They're not. They could be though, because concurrency.